src/uu/paste/src/paste.rs

```diff
@@ -181,32 +270,44 @@ fn paste(
                         Err(e) => return Err(e.map_err_context(String::new)),
                     }
                 }
-                // a buffer of length four is large enough to encode any char
-                let mut buffer = [0; 4];
-                let ch = delimiters[delim_count % delimiters.len()].encode_utf8(&mut buffer);
-                delim_length = ch.len();
 
-                for byte in buffer.iter().take(delim_length) {
-                    output.push(*byte);
+                // Write delimiter, if one exists, to output
+                if let Some(current_delimiter) = delimiter_to_use_option {
+                    output.extend_from_slice(current_delimiter);
                 }
-
-                delim_count += 1;
             }
+
             if files.len() == eof_count {
                 break;
             }
-            // Remove final delimiter
-            output.truncate(output.len() - delim_length);
+
+            if let &mut Some(ref mut de) = &mut delimiter_data_option {
+                let &mut DelimiterData {
+                    current_delimiter_length,
+                    delimiters_encoded,
+                    ref mut delimiters_encoded_iter,
+                } = de;
+
+                // Reset iter after file is processed
+                *delimiters_encoded_iter = delimiters_encoded.iter();
+
+                // Remove trailing delimiter, if there is a delimiter
+                if let Some(us) = output.len().checked_sub(current_delimiter_length) {
+                    output.truncate(us);
+                } else {
+                    // Subtraction would have resulted in a negative number. This should never happen.
```

Comment:
Here it might make sense to use `unreachable!`

Reply:
It ended up being the case that this branch wasn't unreachable. There's a bug in the current implementation (I added a test case for it).

https://github.com/uutils/coreutils/blob/6d79b35dbae1ed5ef001e618b4ac47d5e68915cc/src/uu/paste/src/paste.rs#L153

This is currently overflowing (in some edge cases), but that's only observable in non-release builds.